bin/ch_fuse.c

```diff
@@ -0,0 +1,157 @@
+#include <dirent.h>
+#include <sys/wait.h>
+#include <ll.h>
+
+#include "ch_fuse.h"
+#include "ch_core.h"
+#include "ch_misc.h"
+
+/** Macros **/
+
+/** Types **/
+struct squash {
+   char *mountdir;      // mount location of sqfs
+   pid_t pid;           // ID of fuse loop
+   sqfs_ll_chan chan;   // fuse channel associated with squash fuse session
+   sqfs_ll *ll;         // squashfs image
+};
+
+/** Constants **/
+
+/** Global variables **/
+struct squash sq;
+
+/** Function prototypes (private) **/
+void get_fuse_ops(struct fuse_lowlevel_ops *sqfs_ll_ops);
+void kill_fuse();
+
+/** Functions **/
+
+/* Initalize fuse loop */
+void fuse_loop_init()
+{
+   Ze((fuse_set_signal_handlers(sq.chan.session) == -1), "failed to set signal handlers");
+   if ((fork()) != 0) { //parent process
+      sq.pid = getpid();
+      Ze((fuse_session_loop(sq.chan.session) == -1), "failed to create fuse loop");
+      exit(0);
+   }
+}
+
+/* Function called to start sqfs_ll_clean */
+void kill_fuse()
+{
+   DEBUG("end fuse loop: %d", sq.pid);
+   exit(0);
+}
+
+/* Exit handler for sqfs */
+void sqfs_ll_clean()
+{
+   fuse_remove_signal_handlers(sq.chan.session);
+   sqfs_ll_destroy(sq.ll);
+   DEBUG("unmounting: %s", sq.mountdir);
+   sqfs_ll_unmount(&sq.chan, sq.mountdir);
+   Ze(rmdir(sq.mountdir) == -1, "unable to remove directory");
+}
+
+/* Path is a sqfs*/
+bool sqfs_ll_check(const char *path)
+{
+   sqfs_ll *ll;
+   sqfs_fd_t fd;
+
+   ll = malloc(sizeof(*ll));
+   Tf (ll, "can't allocate memory");
+   memset(ll, 0 , sizeof(*ll));
+   fd = open(path, O_RDONLY);
+   if(fd != -1 && sqfs_init(&ll->fs, fd, 0) == SQFS_OK)
+      return true;
+   sqfs_destroy(&ll->fs);
+   free(ll);
+   return false;
+}
+
+/* Run user command with an extra process for sqfs */
+void sqfs_run_user_command(char *argv[], const char *initial_dir)
+{
+   int status;
+   if (fork() == 0)
+      run_user_command(argv, initial_dir);
+   wait(&status);
+   kill(sq.pid, SIGINT);
+   _Exit(0);
+
+}
+
+/* mounts sqfs image */
+char *sqfs_mount(char *mountdir, char *filepath)
+{
+   //set mountdir to "mountdir/filepath"
+   char *filename, *buffer;
+   path_split(filepath, &buffer, &filename); //get file name
+   split(&filename, &buffer, filename, '.'); //removes first '.'
+   sq.mountdir = cat(cat(mountdir, "/"), filename);
+   DEBUG("mount path: %s", sq.mountdir);
+
+   //init fuse,etc.
+   struct fuse_args args;
+   struct fuse_lowlevel_ops sqfs_ll_ops;
+   args.argc = 1;
+   args.argv = &filepath;
+   args.allocated = 0;
+   get_fuse_ops(&sqfs_ll_ops);
+
+   // mount sqfs
+   sq.ll = sqfs_ll_open(filepath, 0);
+   Te(sq.ll, "%s does not exist", filepath);
+   Ze(opendir(sq.mountdir), "%s already exists", sq.mountdir);
+   Ze(mkdir(sq.mountdir, 0777), "failed to create: %s", sq.mountdir);
+   Te(SQFS_OK == sqfs_ll_mount(&sq.chan, sq.mountdir, &args, &sqfs_ll_ops, sizeof(sqfs_ll_ops), sq.ll), "failed to mount");
+   Ze((sq.chan.session == NULL), "failed to create fuse session");
```

Comment:
These need to pass through the diagnostics from libsquashfuse if at all possible on failure.

Reply:
So there isn't any information from  libsquashfuse for the failure besides the two lines above. FUSE does log errors but it looks like there are all set to error so they will all be printed.

Reply:
👍 If you know how to trigger a failure in `sqfs_ll_mount` then testing the presence of the error message might be a good test.